File: src/main/java/dev/psygamer/wireframe/item/Item.java
package dev.psygamer.wireframe.item;

import dev.psygamer.wireframe.block.state.BlockState;
import dev.psygamer.wireframe.entity.LivingEntity;
import dev.psygamer.wireframe.entity.Player;
import dev.psygamer.wireframe.internal.item.InternalItem;
import dev.psygamer.wireframe.item.util.ClickResult;
import dev.psygamer.wireframe.item.util.ClickResultContainer;
import dev.psygamer.wireframe.item.util.Hand;
import dev.psygamer.wireframe.registry.ItemRegistry;
import dev.psygamer.wireframe.util.BlockPosition;
import dev.psygamer.wireframe.util.Identifier;
import dev.psygamer.wireframe.util.math.BlockHitResult;
import dev.psygamer.wireframe.world.World;

public class Item {
	
	protected final InternalItem internal;
	
	protected final Identifier identifier;
	
	protected final ItemAttributes itemAttributes;
	
	public Item(final Identifier identifier, final ItemAttributes itemAttributes) {
		this.identifier = identifier;
		this.itemAttributes = itemAttributes;
		
		this.internal = new InternalItem(this, itemAttributes);
		
		ItemRegistry.register(this);
	}
	
	public ClickResultContainer<ItemStack> onItemUsed(
			final ItemStack usedItemStack,
			final World world, final Player player, final Hand hand
	) {
		return ClickResultContainer.pass(player.getHeldItem(hand));
	}
	
	public ClickResult onItemUsedOnBlock(
			final ItemStack usedItemStack,
			final World world, final Player player, final Hand hand, final BlockHitResult hitResult
	) {
		return ClickResult.PASS;
	}
	
	public ClickResult onItemUsedOnEntity(
			final ItemStack usedItemStack,
			final World world, final Player player, final LivingEntity entity, final Hand hand
	) {
		return ClickResult.PASS;
	}
	
	public boolean onEntityAttacked(
			final ItemStack usedItemStack,
			final World world, final LivingEntity attacker, final LivingEntity entity
	) {
		return false;
	}
	
	public boolean onBlockMined(
			final ItemStack usedItemStack,
			final BlockState blockState,
			final BlockPosition blockPosition, final World world, final LivingEntity entity
	) {
		return false;
	}
	
	public boolean onItemCrafted(
			final ItemStack usedItemStack, final World world, final Player player
	) {
		return false;
	}
	
	public Identifier getIdentifier() {
		return this.identifier;
	}
	
	public ItemAttributes getItemAttributes() {
		return this.itemAttributes;
	}
	
	public net.minecraft.item.Item getInternal() {
		return this.internal;
	}
}

File: src/main/kotlin/dev/psygamer/wireframe/item/Item.kt
package dev.psygamer.wireframe.item

import dev.psygamer.wireframe.block.state.BlockState
import dev.psygamer.wireframe.entity.LivingEntity
import dev.psygamer.wireframe.entity.Player
import dev.psygamer.wireframe.internal.item.InternalItem
import dev.psygamer.wireframe.item.util.ClickResult
import dev.psygamer.wireframe.item.util.ClickResultContainer
import dev.psygamer.wireframe.item.util.Hand
import dev.psygamer.wireframe.registry.ItemRegistry
import dev.psygamer.wireframe.util.BlockPosition
import dev.psygamer.wireframe.util.Identifier
import dev.psygamer.wireframe.util.math.BlockHitResult
import dev.psygamer.wireframe.wfWrapped
import dev.psygamer.wireframe.world.World

open class Item {
	
	internal val mcNative: net.minecraft.item.Item
	
	val identifier: Identifier
	val itemAttributes: ItemAttributes?
	
	internal constructor(internal: net.minecraft.item.Item) {
		this.identifier = internal.registryName!!.wfWrapped
		this.itemAttributes = null
		
		this.mcNative = internal
	}
	
	protected constructor(identifier: Identifier, itemAttributes: ItemAttributes?) {
		this.identifier = identifier
		this.itemAttributes = itemAttributes
		
		this.mcNative = InternalItem(this, itemAttributes)
		
		ItemRegistry.register(this)
	}
	
	fun onItemUsed(
		itemStack: ItemStack, player: Player, hand: Hand,
		world: World
	): ClickResultContainer<ItemStack> {
		return ClickResultContainer.pass(player.getHeldItem(hand))
	}
	
	open fun onItemUsedOnBlock(
		itemStack: ItemStack, player: Player, hand: Hand,
		world: World, hitResult: BlockHitResult
	): ClickResult {
		return ClickResult.PASS
	}
	
	fun onItemUsedOnEntity(
		itemStack: ItemStack, player: Player, hand: Hand,
		world: World, entity: LivingEntity,
	): ClickResult {
		return ClickResult.PASS
	}
	
	fun onEntityAttacked(
		itemStack: ItemStack, entity: LivingEntity,
		world: World, attacker: LivingEntity,
	): Boolean {
		return false
	}
	
	fun onBlockMined(
		itemStack: ItemStack, entity: LivingEntity,
		world: World, blockPosition: BlockPosition, blockState: BlockState,
	): Boolean {
		return false
	}
	
	fun onItemCrafted(
		usedItemStack: ItemStack, player: Player,
		world: World
	): Boolean {
		return false
	}
}